filebeat/input/filestream/internal/input-logfile/store.go

```diff
@@ -211,6 +211,58 @@ func (s *sourceStore) CleanIf(pred func(v Value) bool) {
 	}
 }
 
+// FixUpIdentifiers copies an existing resource to a new ID and marks the previous one
+// for removal. This method does not check whether the resource belongs to this store
+// beause it is meant to migrate inputs without IDs to inputs with IDs without
+//  duplicating data.
+func (s *sourceStore) FixUpIdentifiers(getNewID func(v Value) (string, interface{})) {
+	s.store.ephemeralStore.mu.Lock()
+	defer s.store.ephemeralStore.mu.Unlock()
+
+	fmt.Println("##################### FixUpIdentifiers prefix", s.identifier.prefix)
+	for key, res := range s.store.ephemeralStore.table {
+		fmt.Println("##################### FixUpIdentifiers key", key)
+		if !s.identifier.MatchesInput(key) {
+			fmt.Println("##################### FixUpIdentifiers, skipping", key)
+			continue
+		}
+
+		if !res.lock.TryLock() {
+			fmt.Println("##################### FixUpIdentifiers cannot lock", key)
+			continue
+		}
+
+		newKey, updatedMeta := getNewID(res)
+		if len(newKey) > 0 && res.internalState.TTL > 0 {
+			if _, ok := s.store.ephemeralStore.table[newKey]; ok {
+				res.lock.Unlock()
```

Comment:
I don't understand this code very well, but calling `Unlock` before a `continue` here and then again when the function returns below seems suspicious to me.

Reply:
The idea here is if the `newKey` is already present on `s.store.ephemeralStore.table` we won't replace it, so we unlock the resource and move on to the next iteration of the `for` loop.

If `newKey` is not present on `s.store.ephemeralStore.table`, then we add it and then unlock at the very end of the `for` loop iteration.

Reply:
I just misread the code here, thanks for clarifying.